test/test_action_builder/test_action_builder.py

```diff
@@ -0,0 +1,345 @@
+    def test_get_tag_by_name(self, m):
+        m.get(
+            f"{self.api_url}/tags?filter=name eq '{self.fake_tag_1}'",
+            text=json.dumps(self.fake_tag_name_search_result),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=2&per_page=25&filter=name eq '{self.fake_tag_1}'",
+            text=json.dumps({"_embedded": {"osdi:tags": []}}),
+        )
+        assert_matching_tables(
+            self.bldr.get_tag_by_name(self.fake_tag_1),
+            Table([self.fake_tags_list_1["_embedded"]["osdi:tags"][0]]),
+        )
+
+    def compare_to_incoming(self, dict1, dict2):
+        # Internal method to compare a reference dict to a new incoming one, keeping only common
+        # keys (or where incoming key is substring). Recursive to handle nesting/JSON.
+
+        incoming = {[key for key in dict1 if k in key][0]: v for k, v in dict2.items()}
+        compare = {k: v for k, v in dict1.items() if k in incoming}
+
+        for key, val in incoming.items():
+            print(key)
+            if isinstance(val, dict):
+                insub, compsub = self.compare_to_incoming(compare[key], val)
+                incoming[key] = insub
+                compare[key] = compsub
+            elif isinstance(val, list) and isinstance(val[0], dict):
+                for i in range(len(val)):
+                    print(i)
+                    insub, compsub = self.compare_to_incoming(compare[key][i], val[i])
+                    incoming[key][i] = insub
+                    compare[key][i] = compsub
+
+        return incoming, compare
+
+    @requests_mock.Mocker()
+    def test_upsert_entity(self, m):
+        m.post(f"{self.api_url}/people", text=json.dumps(self.fake_upserted_response))
+
+        # Flatten and remove items added for spreadable arguments
+        insert_flattened = {
+            **{k: v for k, v in self.fake_insert_person.items() if k != "data"},
+            **self.fake_insert_person["data"]["person"],
+        }
+        insert_incoming, insert_compare = self.compare_to_incoming(
+            self.bldr.upsert_entity(**self.fake_insert_person), insert_flattened
+        )
+
+        update_flattened = {
+            **{k: v for k, v in self.fake_update_person.items() if k != "data"},
+            **self.fake_update_person["data"]["person"],
+        }
+        update_incoming, update_compare = self.compare_to_incoming(
+            self.bldr.upsert_entity(**self.fake_update_person), update_flattened
+        )
+        self.assertEqual(insert_incoming, insert_compare)
+        self.assertEqual(update_incoming, update_compare)
+
+    def tagging_callback(self, request, context):
+        # Internal method for returning the constructed tag data to test
+
+        post_data = request.json()
+        tagging_data = post_data["add_tags"]
+        return tagging_data
+
+    @requests_mock.Mocker()
+    def test_add_tags_to_record(self, m):
+        m.post(f"{self.api_url}/people", json=self.tagging_callback)
```

Comment:
interesting! I've not used requests_mock. is this mocking the json method? what is the `request` param that the `tagging_callback` is receiving?

Reply:
I'm also just wrapping my head around it. The idea as I understand it is to intercept specific HTTP requests and force a desired response. This is done by setting any of the properties available in a `requests` response object. So you'll notice that most of the tests above just return a `text`, and to fit that expectation, it's `json.dumps()` applied to whatever Python object was prepared for the purpose of that test.

In more complicated cases such as this test, we can pass `request` (the details of the request) and `context` (the details of the response) to a callback function, to return something that requires more processing. Specifically here, we're extracting just the list of dicts that should appear under the `add_tags` key of the top-most dict in the request data, and returning that to compare to what we expect that tagging data to look like.